server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,449 @@
+                clearVotingConfigExclusionAndUpdateStatus(false);
+            }
+        });
+    }
+
+    private void clearVotingConfigExclusionAndUpdateStatus(boolean decommissionSuccessful) {
+        decommissionController.clearVotingConfigExclusion(new ActionListener<Void>() {
+            @Override
+            public void onResponse(Void unused) {
+                logger.info(
+                    "successfully cleared voting config exclusion after completing decommission action, proceeding to update metadata"
+                );
+                DecommissionStatus updateStatusWith = decommissionSuccessful ? DecommissionStatus.SUCCESSFUL : DecommissionStatus.FAILED;
+                decommissionController.updateMetadataWithDecommissionStatus(updateStatusWith, statusUpdateListener());
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug(
+                    new ParameterizedMessage("failure in clearing voting config exclusion after processing decommission request"),
+                    e
+                );
+                decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.FAILED, statusUpdateListener());
+            }
+        });
+    }
+
+    private Set<DiscoveryNode> filterNodesWithDecommissionAttribute(
+        ClusterState clusterState,
+        DecommissionAttribute decommissionAttribute,
+        boolean onlyClusterManagerNodes
+    ) {
+        Set<DiscoveryNode> nodesWithDecommissionAttribute = new HashSet<>();
+        Iterator<DiscoveryNode> nodesIter = onlyClusterManagerNodes
+            ? clusterState.nodes().getClusterManagerNodes().valuesIt()
+            : clusterState.nodes().getNodes().valuesIt();
+
+        while (nodesIter.hasNext()) {
+            final DiscoveryNode node = nodesIter.next();
+            if (nodeHasDecommissionedAttribute(node, decommissionAttribute)) {
+                nodesWithDecommissionAttribute.add(node);
+            }
+        }
+        return nodesWithDecommissionAttribute;
+    }
+
+    private static boolean nodeHasDecommissionedAttribute(DiscoveryNode discoveryNode, DecommissionAttribute decommissionAttribute) {
+        return discoveryNode.getAttributes().get(decommissionAttribute.attributeName()).equals(decommissionAttribute.attributeValue());
+    }
+
+    private static void validateAwarenessAttribute(
+        final DecommissionAttribute decommissionAttribute,
+        List<String> awarenessAttributes,
+        Map<String, List<String>> forcedAwarenessAttributes
+    ) {
+        String msg = null;
+        if (awarenessAttributes == null) {
+            msg = "awareness attribute not set to the cluster.";
+        } else if (forcedAwarenessAttributes == null) {
+            msg = "forced awareness attribute not set to the cluster.";
+        } else if (!awarenessAttributes.contains(decommissionAttribute.attributeName())) {
```

Comment:
nit: let's avoid ! and use explicit == false instead

Reply:
done